Support writing integration documentation in markdown

umairwaheed opened the pull request:
Fixes #4011

umairwaheed commented:
@timabbott, @showell, ready for review.

timabbott commented:
This seems to not work for me?

![image](https://cloud.githubusercontent.com/assets/2746074/24673683/fbaf46c0-192d-11e7-9a43-84e922c0d515.png)

timabbott commented:
Also, maybe we should just auto-detect based on which file exists, rather than needing to configure the filename `doc.md`?

umairwaheed commented:
>This seems to not work for me?

Oops. I think I removed the check for existence of `help_content`. Fixed.

>Also, maybe we should just auto-detect based on which file exists, rather than needing to configure the filename doc.md?

Done.

Pushing in a bit.

umairwaheed commented:
@timabbott, fixed now. Please review.

timabbott commented:
Everything looks good except the last commit, which isn't translated faithfully.  E.g. HTML escape sequences aren't correctly replaced with the corresponding character:

![image](https://cloud.githubusercontent.com/assets/2746074/24720200/1d5c5132-19f1-11e7-815e-ef1a97bd1bc5.png)

Can you fix that?  A good technique is to compare side-by-side the docs on chat.zulip.org with on your branch, so you can easily look for formatting differences.

timabbott commented on zerver/lib/integrations.py:
```diff
@@ -74,14 +84,30 @@ def __init__(self, name, client_name=None, logo=None, secondary_line_text=None,
+                doc = self.DEFAULT_DOC_PATH.format(name=name, ext='md')
+                loader.get_template(doc)
+            except loader.TemplateDoesNotExist:
+                doc = self.DEFAULT_DOC_PATH.format(name=name, ext='html')
```
Actually, I have a concern here -- since this runs at import time, is `loader.get_template()` expensive?  If so, we might want to do something cheaper.  

I did some quick testing and confirmed that this appears to add like 10ms to the import time for this module:
```
# Post-branch
In [2]: print time.time(); import zerver.lib.integrations; print time.time()
1491416444.57
1491416444.7

# Pre-branch
In [2]: print time.time(); import zerver.lib.integrations; print time.time()
1491416460.92
1491416460.98
```

Should we just use an `os.path.join()` check to see which file exists rather than `loader.get_template()`?

umairwaheed replied:
Using `os.path.exists` now.

timabbott commented:
FYI @umairwaheed I just merged the fix for #4293, which will merge conflict with this a bit.

umairwaheed commented:
@timabbott, updated. Please review.

timabbott commented:
Merged, thanks @umairwaheed!
